check: add a new linter that checks that quoted URLs are live.

knz opened the pull request:
This was requested by @sploiselle in https://github.com/cockroachdb/cockroach/pull/13540#issuecomment-279493990

Fixes #15752.
Fixes  #15844.

knz committed:
check: add a new linter that checks that quoted URLs are live.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15751)
<!-- Reviewable:end -->

knz commented:
This was requested by @sploiselle in https://github.com/cockroachdb/cockroach/pull/13540#issuecomment-279493990

knz commented:
I acknowledge that this may better be served by a nightly test instead of a linter. Suggestions / opinions welcome.

tamird commented:
Awesome! Did you mean to include all the fixes in this commit, too?

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[build/style_test.go, line 216 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZtv5RpNbtl2C8zRsL:-KjZtv5RpNbtl2C8zRsM:bi42y3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L216)):*
> ```Go
> 	t.Run("TestURL", func(t *testing.T) {
> 		if testing.Short() {
> 			t.Skip("short flag")
> ```

How long does this test take to run?

---

*[build/style_test.go, line 220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZvYX4_jsJq7bfyXDx:-KjZvYX4_jsJq7bfyXDy:b-hpmcf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L220)):*
> ```Go
> 
> 		t.Parallel()
> 		const urlRE = `https?://[^ \t\n/$.?#].[^ \t\n)]*(\)\.aspx)?`
> ```

where did this regexp come from? I found one [here](http://stackoverflow.com/questions/6038061/regular-expression-to-find-urls-within-a-string): `(http|ftp|https)://([\w_-]+(?:(?:\.[\w_-]+)+))([\w.,@?^=%&:/~+#-]*[\w@?^=%&/~+#-])?`

---

*[build/style_test.go, line 221 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZveQEvZwimjtBQ2d1:-KjZveQEvZwimjtBQ2d2:b-ujwm6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L221)):*
> ```Go
> 		t.Parallel()
> 		const urlRE = `https?://[^ \t\n/$.?#].[^ \t\n)]*(\)\.aspx)?`
> 		re := regexp.MustCompile(urlRE)
> ```

can you hoist this out so it panics at init time?

---

*[build/style_test.go, line 223 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZuP14Xjd8bIzSu5f_:-KjZuP14Xjd8bIzSu5fa:bhbp5ea) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L223)):*
> ```Go
> 		re := regexp.MustCompile(urlRE)
> 
> 		cmd, stderr, filter, err := dirCmd(pkg.Dir, "git", "grep", "-nE", urlRE)
> ```

I think you want `-oE` here, or `-noE` to avoid the extra regexp matching you do in the `ForEach` below.

---

*[build/style_test.go, line 255 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZuuD79ur_TIHk427h:-KjZuuD79ur_TIHk427i:b-1et9rx) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L255)):*
> ```Go
> 		}
> 
> 		ignored := []string{
> ```

nit: consider making this a `map[string]struct{}`

---

*[build/style_test.go, line 264 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZv-UzlA6wdYzGYV8f:-KjZv-UzlA6wdYzGYV8g:bexwdn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L264)):*
> ```Go
> 			"http://localhost",
> 			"http://metadata/",
> 			"http://www.bohemiancoding.com/sketch/ns",
> ```

a bunch of these look like legit websites, why are they ignored? we should have comments here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-:-KjZtyk-T9KP7RPuQCjP:brwva7d)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Pretty cool! How long does it take to run? We may want to issue multiple HTTP requests in parallel (could move the inner part of the loop inside a `go func() {}` and use a semaphore).

tschottdorf commented:
Pretty cool, but is it OK to flood all the URLs on any CI run or local `make check` invocation? And I often run `make check` when I'm offline (train, plane, ...) so would be nice not to have it fail there (i.e. perhaps it should be a package-level test somewhere?) Those are just suggestions, I'm glad that we're trapping dead links. Feel free to discard.

dt commented:
+1 on making this only a nightly job, as opposed to every `make` invocation, particularly since the URLs can rot on their own too and thus might be unrelated to any local changes being checked.

I'd either use an env var to skip or a build tag, and then add a nightly CI to pass said var/tag.

We could also use a whitelist of prefixes instead of a blacklist, and allow optionally remapping those prefixes, e.g. so a docs build could to its staged, local URL and run the checker to make sure that the docs changes don't break any links either.

knz committed:
*: fix stale URLs

knz commented:
@dt I added a build tag. I'll add a separate TC nightly script if/when this merges.
Note this is not only about catching doc changes. 

> @tamird Did you mean to include all the fixes in this commit, too?

ok done

(Now tyring to add some HTTP concurrency as suggested by Radu)

---

Review status: 0 of 37 files reviewed at latest revision, 6 unresolved discussions.

---

*[build/style_test.go, line 216 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZtv5RpNbtl2C8zRsL:-KjaVMlWGo35crDkTVKC:bmag5tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L216)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

How long does this test take to run?
</blockquote></details>

~90 seconds. But foremost I am concerned about flooding all these URLs  on every `make check` run. I did like @dt suggested and added a build tag instead, so we only run this on nightlies / explicitly.

---

*[build/style_test.go, line 220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZvYX4_jsJq7bfyXDx:-KjaVXrdbZC22_ra7acN:b-g7ftvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L220)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

where did this regexp come from? I found one [here](http://stackoverflow.com/questions/6038061/regular-expression-to-find-urls-within-a-string): `(http|ftp|https)://([\w_-]+(?:(?:\.[\w_-]+)+))([\w.,@?^=%&:/~+#-]*[\w@?^=%&/~+#-])?`
</blockquote></details>

Added a reference. Yours didn't work for me.

---

*[build/style_test.go, line 221 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZveQEvZwimjtBQ2d1:-KjaVb9z26AQrFkzbmNx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L221)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you hoist this out so it panics at init time?
</blockquote></details>

Done.

---

*[build/style_test.go, line 223 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZuP14Xjd8bIzSu5f_:-KjaVqxNgzTRaqbgNpHp:bwwmv3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L223)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think you want `-oE` here, or `-noE` to avoid the extra regexp matching you do in the `ForEach` below.
</blockquote></details>

`git grep` does not support -o (only `grep` does)

---

*[build/style_test.go, line 255 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZuuD79ur_TIHk427h:-KjaVvt_gvTvIGtXVKjn:b-n7qd3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L255)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: consider making this a `map[string]struct{}`
</blockquote></details>

I'm not sure how that helps. The match has to be done with HasPrefix; if I have to scan the entire data structure anyway, an array is faster.

---

*[build/style_test.go, line 264 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZv-UzlA6wdYzGYV8f:-KjaWKagheYUbCwmIoYH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L264)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

a bunch of these look like legit websites, why are they ignored? we should have comments here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-:-KjaWP5V862EoSSNFErz:b-bt6j36)*
<!-- Sent from Reviewable.io -->

knz commented:
@RaduBerinde: Done.

---

Review status: 0 of 38 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-:-KjacwJlK1JoBqL6Q32V:b-bucfh0)*
<!-- Sent from Reviewable.io -->

dt commented on build/url_test.go:
```diff
@@ -0,0 +1,174 @@
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+	client := &http.Client{Transport: tr}
```
`Timeout: time.Second * 3,`?

dt commented on build/url_test.go:
```diff
@@ -0,0 +1,174 @@
+			if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+				if resp.StatusCode == 405 {
+					// The server doesn't like HEAD. Try GET.
+					resp, err = client.Get(url)
```
I think you need to fully read and then Close resp.Body to make the client happy, otherwise the underlying connection isn't freed for reuse.

Copy(ioutil.Discard, resp.Body) && resp.Body.Close if memory serves

dt commented on build/style_test.go:
```diff
@@ -803,4 +803,6 @@ func TestStyle(t *testing.T) {
 		}
 	})
+
+	t.Run("TestURLLiveness", func(t *testing.T) { testURLLiveness(t, pkg) })
```
Could just make it a 1st class test and avoid the empty stub in the common case?

knz commented:
Review status: 0 of 38 files reviewed at latest revision, 9 unresolved discussions.

---

*[build/style_test.go, line 807 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZse03x5Ocif05taiH-r3-807:-Kjc3EKhYzFGSSmMxamg:bo9f2c8) ([raw file](https://github.com/cockroachdb/cockroach/blob/36fd4b521a92ecc8bc99429b1d4efed1c4b6e17f/build/style_test.go#L807)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Could just make it a 1st class test and avoid the empty stub in the common case?
</blockquote></details>

That was my first move, but then you can't do `make check TESTS=TestURLLiveness TAGS=urlcheck` any more (`make check` wants all the things to be sub-tests of TestStyle)

---

*[build/url_test.go, line 124 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjaZ0kcjGmJpkUbDjDv-r3-124:-Kjc2Qb8uZOckNNVPHPo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36fd4b521a92ecc8bc99429b1d4efed1c4b6e17f/build/url_test.go#L124)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

`Timeout: time.Second * 3,`?
</blockquote></details>

Done.

---

*[build/url_test.go, line 158 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjaZ0kcjGmJpkUbDjDv-r3-158:-Kjc3Cglsz3zi-snV2O2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36fd4b521a92ecc8bc99429b1d4efed1c4b6e17f/build/url_test.go#L158)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think you need to fully read and then Close resp.Body to make the client happy, otherwise the underlying connection isn't freed for reuse.

Copy(ioutil.Discard, resp.Body) && resp.Body.Close if memory serves
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 38 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[build/style_test.go, line 807 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZse03x5Ocif05taiH-r3-807:-KjccYB0uTQ6mF8pmVSM:b4rs41c) ([raw file](https://github.com/cockroachdb/cockroach/blob/36fd4b521a92ecc8bc99429b1d4efed1c4b6e17f/build/style_test.go#L807)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

That was my first move, but then you can't do `make check TESTS=TestURLLiveness TAGS=urlcheck` any more (`make check` wants all the things to be sub-tests of TestStyle)
</blockquote></details>

yeah, I guess I suggesting it just not be in `make check` but rather be standalone.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 38 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[build/style_test.go, line 807 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZse03x5Ocif05taiH-r3-807:-KjcsKb7mvxrYYiLTRw4:b-rharjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/36fd4b521a92ecc8bc99429b1d4efed1c4b6e17f/build/style_test.go#L807)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

yeah, I guess I suggesting it just not be in `make check` but rather be standalone.
</blockquote></details>

I don't care either way. @tamird you have an opinion?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751)*
<!-- Sent from Reviewable.io -->

knz commented:
Also separate issue, can someone help me figure out what's wrong in TC?

```
style_test.go:410: exec: "crlfmt": executable file not found in $PATH
```

dt commented:
running in builder and bootstrapped?

tamird commented:
Where is the urlcheck tag coming from? I think we're going to want to make this a command called from a new `teamcity-urlcheck.sh` script.

---

Reviewed 36 of 36 files at r2, 2 of 3 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[build/style_test.go, line 223 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZuP14Xjd8bIzSu5f_:-KjdYd8a6Lrghb5GBWTG:bsrrgsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L223)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`git grep` does not support -o (only `grep` does)
</blockquote></details>

Oh dear :crying_cat_face: 

---

*[build/style_test.go, line 255 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZuuD79ur_TIHk427h:-KjdYg_0WTeinuNsM12S:bv0zz7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/319b5a18d0094825aa1ada87d7a25d6926ede636/build/style_test.go#L255)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'm not sure how that helps. The match has to be done with HasPrefix; if I have to scan the entire data structure anyway, an array is faster.
</blockquote></details>

You're correct, my mistake!

---

*[build/style_test.go, line 807 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZse03x5Ocif05taiH-r3-807:-KjdYk8_IhIQu5zosivj:bstwg4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/36fd4b521a92ecc8bc99429b1d4efed1c4b6e17f/build/style_test.go#L807)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't care either way. @tamird you have an opinion?
</blockquote></details>

Yeah, I'm with @dt in that I'd prefer a standalone thing.

---

*[build/url_test.go, line 37 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-Kjd_vZVzyGlAojIEw4G:-Kjd_vZVzyGlAojIEw4H:buy2sa4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b957e5b1b4ef15bed0bab21639dc6229cd74140/build/url_test.go#L37)):*
> ```Go
> )
> 
> // Number of concurrent HTTP requests during the test
> ```

is this actually necessary? I'm pretty sure it's safe to let net/http handle this.

---

*[build/url_test.go, line 61 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjdbHVyhCNzCDVPYBbe:-KjdbHVzu7W6wVvYdJ5K:bekna8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b957e5b1b4ef15bed0bab21639dc6229cd74140/build/url_test.go#L61)):*
> ```Go
> 	"https://github.com/cockroachdb/cockroach/commits/%s",
> 	"https://github.com/cockroachdb/cockroach/issues/%d",
> 	// These are AWS endpoints.
> ```

I believe metadata is a GCE endpoint, but I think this can just say "these are internal to public cloud providers"

---

*[build/url_test.go, line 65 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjdawC35yStQn9ZLk3r:-KjdawC35yStQn9ZLk3s:b-b6n1ws) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b957e5b1b4ef15bed0bab21639dc6229cd74140/build/url_test.go#L65)):*
Really? I just tested one of these and got a 200 OK:
```
$ curl -sfSL -I -v https://registry.yarnpkg.com/@types/mocha/-/mocha-2.2.40.tgz#9811dd800ece544cd84b5b859917bf584a150c4c
*   Trying 104.16.63.173...
* TCP_NODELAY set
* Connected to registry.yarnpkg.com (104.16.63.173) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
* Server certificate: ssl462304.cloudflaressl.com
* Server certificate: COMODO ECC Domain Validation Secure Server CA 2
* Server certificate: COMODO ECC Certification Authority
> HEAD /@types/mocha/-/mocha-2.2.40.tgz HTTP/1.1
> Host: registry.yarnpkg.com
> User-Agent: curl/7.51.0
> Accept: */*
>
< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< Date: Mon, 08 May 2017 19:50:25 GMT
Date: Mon, 08 May 2017 19:50:25 GMT
< Content-Type: application/octet-stream
Content-Type: application/octet-stream
< Content-Length: 2672
Content-Length: 2672
< Connection: keep-alive
Connection: keep-alive
< Set-Cookie: __cfduid=d8d9bab785c6bed1cbf834717649a09921494273024; expires=Tue, 08-May-18 19:50:24 GMT; path=/; domain=.yarnpkg.com; HttpOnly
Set-Cookie: __cfduid=d8d9bab785c6bed1cbf834717649a09921494273024; expires=Tue, 08-May-18 19:50:24 GMT; path=/; domain=.yarnpkg.com; HttpOnly
< last-modified: Sat, 11 Mar 2017 01:36:44 GMT
last-modified: Sat, 11 Mar 2017 01:36:44 GMT
< etag: "58c354ac-a70"
etag: "58c354ac-a70"
< expires: Tue, 09 May 2017 01:50:25 GMT
expires: Tue, 09 May 2017 01:50:25 GMT
< access-control-allow-origin: *
access-control-allow-origin: *
< access-control-allow-methods: GET
access-control-allow-methods: GET
< Cache-Control: max-age=0
Cache-Control: max-age=0
< Accept-Ranges: bytes
Accept-Ranges: bytes
< Via: 1.1 varnish
Via: 1.1 varnish
< Age: 0
Age: 0
< X-Served-By: cache-jfk8147-JFK
X-Served-By: cache-jfk8147-JFK
< X-Cache: MISS
X-Cache: MISS
< X-Cache-Hits: 0
X-Cache-Hits: 0
< X-Timer: S1494273025.948830,VS0,VE113
X-Timer: S1494273025.948830,VS0,VE113
< Server: cloudflare-nginx
Server: cloudflare-nginx
< CF-RAY: 35bef2a5ed5421f2-EWR
CF-RAY: 35bef2a5ed5421f2-EWR

<
* Curl_http_done: called premature == 0
* Connection #0 to host registry.yarnpkg.com left intact
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-:-Kjda7qXgg9rz7N8HHkb:b-st2191)*
<!-- Sent from Reviewable.io -->

knz commented:
Please hold my hand there: 1) where do you want this command to live 2) where should I place its source code 3) what makefile rule do you want me to define 

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[build/style_test.go, line 807 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjZse03x5Ocif05taiH-r3-807:-KjmW5TsQeems15biKQt:b-iobuo6) ([raw file](https://github.com/cockroachdb/cockroach/blob/36fd4b521a92ecc8bc99429b1d4efed1c4b6e17f/build/style_test.go#L807)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Yeah, I'm with @dt in that I'd prefer a standalone thing.
</blockquote></details>

See my comment above

---

*[build/url_test.go, line 37 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-Kjd_vZVzyGlAojIEw4G:-KjmW6aIXRZLT3u_WPH1:b-rkgi6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b957e5b1b4ef15bed0bab21639dc6229cd74140/build/url_test.go#L37)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is this actually necessary? I'm pretty sure it's safe to let net/http handle this.
</blockquote></details>

It doesn't (as far as I can see), but I can be convinced otherwise if you point me to the right location in the http package's sources.

---

*[build/url_test.go, line 61 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjdbHVyhCNzCDVPYBbe:-KjmWGlvOLKrru0xmiAF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b957e5b1b4ef15bed0bab21639dc6229cd74140/build/url_test.go#L61)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I believe metadata is a GCE endpoint, but I think this can just say "these are internal to public cloud providers"
</blockquote></details>

Done.

---

*[build/url_test.go, line 65 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-KjdawC35yStQn9ZLk3r:-KjmWNsWprSU3tZi4f6T:bv0nen8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b957e5b1b4ef15bed0bab21639dc6229cd74140/build/url_test.go#L65)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Really? I just tested one of these and got a 200 OK:
```
$ curl -sfSL -I -v https://registry.yarnpkg.com/@types/mocha/-/mocha-2.2.40.tgz#9811dd800ece544cd84b5b859917bf584a150c4c
*   Trying 104.16.63.173...
* TCP_NODELAY set
* Connected to registry.yarnpkg.com (104.16.63.173) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
* Server certificate: ssl462304.cloudflaressl.com
* Server certificate: COMODO ECC Domain Validation Secure Server CA 2
* Server certificate: COMODO ECC Certification Authority
> HEAD /@types/mocha/-/mocha-2.2.40.tgz HTTP/1.1
> Host: registry.yarnpkg.com
> User-Agent: curl/7.51.0
> Accept: */*
>
< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< Date: Mon, 08 May 2017 19:50:25 GMT
Date: Mon, 08 May 2017 19:50:25 GMT
< Content-Type: application/octet-stream
Content-Type: application/octet-stream
< Content-Length: 2672
Content-Length: 2672
< Connection: keep-alive
Connection: keep-alive
< Set-Cookie: __cfduid=d8d9bab785c6bed1cbf834717649a09921494273024; expires=Tue, 08-May-18 19:50:24 GMT; path=/; domain=.yarnpkg.com; HttpOnly
Set-Cookie: __cfduid=d8d9bab785c6bed1cbf834717649a09921494273024; expires=Tue, 08-May-18 19:50:24 GMT; path=/; domain=.yarnpkg.com; HttpOnly
< last-modified: Sat, 11 Mar 2017 01:36:44 GMT
last-modified: Sat, 11 Mar 2017 01:36:44 GMT
< etag: "58c354ac-a70"
etag: "58c354ac-a70"
< expires: Tue, 09 May 2017 01:50:25 GMT
expires: Tue, 09 May 2017 01:50:25 GMT
< access-control-allow-origin: *
access-control-allow-origin: *
< access-control-allow-methods: GET
access-control-allow-methods: GET
< Cache-Control: max-age=0
Cache-Control: max-age=0
< Accept-Ranges: bytes
Accept-Ranges: bytes
< Via: 1.1 varnish
Via: 1.1 varnish
< Age: 0
Age: 0
< X-Served-By: cache-jfk8147-JFK
X-Served-By: cache-jfk8147-JFK
< X-Cache: MISS
X-Cache: MISS
< X-Cache-Hits: 0
X-Cache-Hits: 0
< X-Timer: S1494273025.948830,VS0,VE113
X-Timer: S1494273025.948830,VS0,VE113
< Server: cloudflare-nginx
Server: cloudflare-nginx
< CF-RAY: 35bef2a5ed5421f2-EWR
CF-RAY: 35bef2a5ed5421f2-EWR

<
* Curl_http_done: called premature == 0
* Connection #0 to host registry.yarnpkg.com left intact
```
</blockquote></details>

Read the comment: "these are **valid** by definition"
It makes no sense to auto-test them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15751#-:-KjmVxaSv9Ff2adQ1q9I:b464qbp)*
<!-- Sent from Reviewable.io -->

knz commented:
Closing because of "bandwidth exceeded" error on my side.

dt commented:
Sorry my comments weren't clear/helpful.

`running in builder and bootstrapped?` was an attempt to help with your TC error since those are things I've missed in the past and have learned to check first when debugging a CI issue, but I was too terse there and was probably wrong anyway. 

If you decide to come back to this at some point, to clarify my suggestion of a standalone test, I was picturing, essentially just capitalizing your `testURLLiveness` helper in`url_test.go` (and inlining that one use of pkg.Dir), making it into a 1st class Go test, then removing the subtest addition in style_test.go.

That test could be run with something like `go test -run TestURLLiveness -tags urlcheck ./build`, (potentially wrapped in a makefile rule?), invoked (using builder.sh) in some sort of `teamcity-urlcheck.sh` script run nightly, on these points @tamird has more specific suggestions than I do.

tamird commented:
I'm happy to take this over on Friday. Let's leave it for now.

On Wed, May 10, 2017 at 10:34 AM, David Taylor <notifications@github.com>
wrote:

> Sorry my comments weren't clear/helpful.
>
> running in builder and bootstrapped? was an attempt to help with your TC
> error since those are things I've missed in the past and have learned to
> check first when debugging a CI issue, but I was too terse there and was
> probably wrong anyway.
>
> If you decide to come back to this at some point, to clarify my suggestion
> of a standalone test, I was picturing, essentially just capitalizing your
> testURLLiveness helper inurl_test.go (and inlining that one use of
> pkg.Dir), making it into a 1st class Go test, then removing the subtest
> addition in style_test.go.
>
> That test could be run with something like go test -run TestURLLiveness
> -tags urlcheck ./build, (potentially wrapped in a makefile rule?),
> invoked (using builder.sh) in some sort of teamcity-urlcheck.sh script
> run nightly, on these points @tamird <https://github.com/tamird> has more
> specific suggestions than I do.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/15751#issuecomment-300501807>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPDqUlkYjWKNhmQFDAB4x_UjSvXZCks5r4csPgaJpZM4NTUCI>
> .
>
